common/src/include/ebbrt/MulticoreEbb.h

```diff
@@ -11,53 +12,161 @@
-        return *it->second;
-      }
-      // we failed to find a rep, we must construct one
-      rep = new T(root);
-    }
 
+  static std::pair<Root*, boost::container::flat_map<size_t, T*>>*
+  GetMapEntry(EbbId id) {
     LocalIdMap::Accessor accessor;
     auto found = local_id_map->Find(accessor, id);
     if (!found)
       throw std::runtime_error("Failed to find root for MulticoreEbb");
 
-    auto pair = boost::any_cast<std::pair<Root*, RepMap>>(&accessor->second);
-    auto& rep_map = pair->second;
-    rep_map[Cpu::GetMine()] = rep;
-    EbbRef<T>::CacheRef(id, *rep);
-    return *rep;
+    return boost::any_cast<
+        std::pair<Root*, boost::container::flat_map<size_t, T*>>>(
+        &accessor->second);
   }
+};
+
+template <class T, class Root = void>
+class MulticoreEbb : public MulticoreEbbBase<T, Root> {
+ public:
+  static T& HandleFault(EbbId id);
 
  private:
   typedef boost::container::flat_map<size_t, T*> RepMap;
 };
 
+template <class T>
+class MulticoreEbb<T, MulticoreEbbRoot<T>>
+    : public MulticoreEbbBase<T, MulticoreEbbRoot<T>> {
+ public:
+  static T& HandleFault(EbbId id);
+  void SetRoot(MulticoreEbbRoot<T>* root) { root_ = root; }
+  const MulticoreEbbRoot<T>* GetRoot() { return root_; }
+
+ private:
+  const MulticoreEbbRoot<T>* root_;
+  typedef boost::container::flat_map<size_t, T*> RepMap;
+};
+
+template <class T> class MulticoreEbb<T, void> {
+ public:
+  static EbbRef<T> Create(EbbId id = ebb_allocator->Allocate());
+  static T& HandleFault(EbbId id);
+
+ private:
+  typedef boost::container::flat_map<size_t, T*> RepMap;
+};
+
+template <class T> EbbRef<T> MulticoreEbb<T, void>::Create(EbbId id) {
+  local_id_map->Insert(std::make_pair(id, RepMap()));
+  return EbbRef<T>(id);
+}
+
+template <class T, class Root> T& MulticoreEbb<T, Root>::HandleFault(EbbId id) {
+  auto pair = MulticoreEbbBase<T, Root>::GetMapEntry(id);
+  const auto& root = *(pair->first);
+  auto rep_map = boost::any_cast<RepMap>(pair->second);
+  auto it = rep_map.find(Cpu::GetMine());
+  if (it != rep_map.end()) {
+    EbbRef<T>::CacheRef(id, *it->second);
+    return *it->second;
+  }
+  // we failed to find a rep, we must construct one
+  T* rep = new T(root);
+  auto& set_map = pair->second;
+  set_map[Cpu::GetMine()] = rep;
+  EbbRef<T>::CacheRef(id, *rep);
+  return *rep;
+}
+
+template <class T>
+T& MulticoreEbb<T, MulticoreEbbRoot<T>>::HandleFault(EbbId id) {
+  auto pair = MulticoreEbbBase<T, MulticoreEbbRoot<T>>::GetMapEntry(id);
+  auto rep_map = boost::any_cast<RepMap>(pair->second);
+  auto it = rep_map.find(Cpu::GetMine());
+  if (it != rep_map.end()) {
+    EbbRef<T>::CacheRef(id, *it->second);
+    return *it->second;
+  }
+  // we failed to find a rep, we must construct one
+  auto root = pair->first;
+  T* rep = new T();
+  root->SetEbbId(id);
+  rep->SetRoot(root);
+  auto& set_map = pair->second;
+  set_map[Cpu::GetMine()] = rep;
+  EbbRef<T>::CacheRef(id, *rep);
+  return *rep;
+}
```

Comment:
Seems like a large amount of this function could be extracted out to reduce repetition